src/ddsrt/src/events/kqueue/events_kqueue.c

```diff
@@ -0,0 +1,388 @@
+  queue->cnewevents = EVENTS_CONTAINER_DELTA;
+
+  /*create kevents array*/
+  queue->kevents = ddsrt_malloc(sizeof(struct kevent) * queue->cevents);
+  if (NULL == queue->kevents)
+    goto alloc_fail_0;
+  queue->events = ddsrt_malloc(sizeof(queued_event_t) * queue->cevents);
+  if (NULL == queue->events)
+    goto alloc_fail_1;
+  queue->newevents = ddsrt_malloc(sizeof(queued_event_t) * queue->cnewevents);
+  if (NULL == queue->newevents)
+    goto alloc_fail_2;
+
+  /*create kevent polling instance */
+  if (-1 == (queue->kq = kqueue()))
+    goto kq_fail;
+  else if (-1 == fcntl(queue->kq, F_SETFD, fcntl(queue->kq, F_GETFD) | FD_CLOEXEC))
+    goto pipe0_fail;
+  /*create interrupt pipe */
+  else if (-1 == pipe(queue->interrupt))
+    goto pipe0_fail;
+  else if (-1 == fcntl(queue->interrupt[0], F_SETFD, fcntl(queue->interrupt[0], F_GETFD) | FD_CLOEXEC) ||
+           -1 == fcntl(queue->interrupt[1], F_SETFD, fcntl(queue->interrupt[1], F_GETFD) | FD_CLOEXEC))
+    goto pipe1_fail;
+
+  /*register interrupt event*/
+  struct kevent kev;
+  EV_SET(&kev, queue->interrupt[0], EVFILT_READ, EV_ADD, 0, 0, 0);
+  if (-1 == kevent(queue->kq, &kev, 1, NULL, 0, NULL))
+    goto pipe1_fail;
+
+  ddsrt_mutex_init(&queue->lock);
+
+  return DDS_RETCODE_OK;
+
+pipe1_fail:
+  close(queue->interrupt[0]);
+  close(queue->interrupt[1]);
+pipe0_fail:
+  close(queue->kq);
+kq_fail:
+  ddsrt_free(queue->newevents);
+alloc_fail_2:
+  ddsrt_free(queue->events);
+alloc_fail_1:
+  ddsrt_free(queue->kevents);
+alloc_fail_0:
+  return DDS_RETCODE_ERROR;
+}
+
+/**
+* @brief Finishes an event queue.
+*
+* Will free created containers and close interrupt pipe and kernel event monitor.
+*
+* @param[in,out] queue The queue to finish.
+*/
+static void ddsrt_event_queue_fini(ddsrt_event_queue_t* queue) ddsrt_nonnull_all;
+
+void ddsrt_event_queue_fini(ddsrt_event_queue_t* queue)
+{
+  close(queue->interrupt[0]);
+  close(queue->interrupt[1]);
+  close(queue->kq);
+
+  ddsrt_mutex_destroy(&queue->lock);
+  ddsrt_free(queue->events);
+  ddsrt_free(queue->kevents);
+  ddsrt_free(queue->newevents);
+}
+
+ddsrt_event_queue_t* ddsrt_event_queue_create(void)
+{
+  ddsrt_event_queue_t* returnptr = ddsrt_malloc(sizeof(ddsrt_event_queue_t));
+  if (DDS_RETCODE_OK != ddsrt_event_queue_init(returnptr)) {
+    ddsrt_free(returnptr);
+    returnptr = NULL;
+  }
+  return returnptr;
+}
+
+void ddsrt_event_queue_delete(ddsrt_event_queue_t* queue)
+{
+  ddsrt_event_queue_fini(queue);
+  ddsrt_free(queue);
+}
+
+size_t ddsrt_event_queue_nevents(ddsrt_event_queue_t* queue)
+{
+  size_t ret;
+  ddsrt_mutex_lock(&queue->lock);
+  ret = queue->nevents + queue->nnewevents - queue->ndeleted;
+  ddsrt_mutex_unlock(&queue->lock);
+  return ret;
+}
+
+dds_return_t ddsrt_event_queue_wait(ddsrt_event_queue_t* queue, dds_duration_t reltime)
+{
+  dds_return_t ret = DDS_RETCODE_OK;
+
+  ddsrt_mutex_lock(&queue->lock);
+
+  /*remove deregistered events*/
+  size_t i = 0;
+  while (i < queue->nevents)
```

Comment:
I will think of some way that this will not be executed every time a wait is called, maybe a flag indicating that the set of events has been modified?

Reply:
This is now implemented.